nav2_collision_monitor/src/collision_monitor_node.cpp

```diff
@@ -0,0 +1,530 @@
+      // Setting up zero velocity for stopping model
+      robot_action.action_type = STOP;
+      robot_action.req_vel.x = 0.0;
+      robot_action.req_vel.y = 0.0;
+      robot_action.req_vel.tw = 0.0;
+    } else {  // SLOWDOWN
+      const Velocity safe_vel = velocity * polygon->getSlowdownRatio();
+      // Check that currently calculated velocity is safer than
+      // chosen for previous shapes one
+      if (robot_action.action_type == DO_NOTHING || safe_vel < robot_action.req_vel) {
+        robot_action.action_type = SLOWDOWN;
+        robot_action.req_vel = safe_vel;
+      }
+    }
+  }
+}
+
+void CollisionMonitor::processApproach(
+    const std::shared_ptr<PolygonBase> polygon,
+    const std::vector<Point> & collision_points,
+    const Velocity & velocity,
+    double & min_collision_time,
+    Action & robot_action)
+{
+  const double collision_time = polygon->getCollisionTime(collision_points, velocity);
+  if (collision_time >= 0.0 &&
+    collision_time <= polygon->getTimeBeforeCollision() &&
+    collision_time < min_collision_time)
+  {
+    const Velocity safe_vel = polygon->getSafeVelocity(velocity, collision_time);
+    // Check that currently calculated velocity is safer than
+    // chosen for previous shapes one
+    if (robot_action.action_type == DO_NOTHING || safe_vel < robot_action.req_vel) {
+      robot_action.action_type = APPROACH;
+      robot_action.req_vel = safe_vel;
+      min_collision_time = collision_time;
+    }
+  }
+}
+
+void CollisionMonitor::releaseOperation(
+  const rclcpp::Time & curr_time, const Velocity & desired_vel, Action & robot_action)
+{
+  Velocity release_vel = robot_action_prev_.req_vel;
+  // Previous robot speed
+  const double prev_vel_module =
+    std::sqrt(release_vel.x * release_vel.x + release_vel.y * release_vel.y);
+  // Speed increase step: acceleration during (curr_time - prev_time_) interval
+  const double vel_change = release_acceleration_ * (curr_time - prev_time_).seconds();
+  // Maximum (desired) robot speed
+  const double desired_vel_module =
+    std::sqrt(desired_vel.x * desired_vel.x + desired_vel.y * desired_vel.y);
+
+  // Calculate speed ratio as previous speed increased on acceleration
+  // during (curr_time - prev_time_) time interval towards desired velocity
+  const double change_ratio = (prev_vel_module + vel_change) / desired_vel_module;
+  release_vel.x = desired_vel.x * change_ratio;
+  release_vel.y = desired_vel.y * change_ratio;
+  release_vel.tw = desired_vel.tw * change_ratio;
+
+  // Check that approximately increased velocity does not exceed the desired speed
+  if (release_vel < desired_vel) {
+    robot_action.req_vel = release_vel;
+  } else {
+    robot_action.req_vel = desired_vel;
+    // Robot already reached its desired speed: continuing normal operation
+    release_operation_ = false;
+  }
+}
+
+void CollisionMonitor::printAction(const Action & robot_action)
+{
+  if (robot_action.action_type == STOP) {
```

Comment:
These logs are not very clear and I'm not sure myself exactly how to read them

Reply:
Why not just say stopping due to `polygon_name` or slowing down N% due to `polygon_name` or within `Ns` of a collision so scaling back velocity.

Reply:
Done by new printAction() routine: https://github.com/ros-planning/navigation2/pull/2982/files#diff-e1c76d88179c14a1cf399e94bebd4f5d17dedca39e09b3bbdc810aa11a705543R573